### PR TITLE
 Introduce API to better control view annotation positioning 

### PR DESCRIPTION
**Pull request by kiryldz**
## Pull request checklist:
 - [ ] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
    
 - [ ] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Optimize code for java consumption (`@JvmOverloads`, `@file:JvmName`, etc).
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [ ] Run `make update-api` to update generated api files, if there's public API changes, otherwise the `verify-api-*` CI steps might fail.
 - [ ] Update [CHANGELOG.md](../CHANGELOG.md) or use the label 'skip changelog', otherwise `check changelog` CI step will fail.
 - [ ] If this PR is a `v10.[version]` release branch fix / enhancement, merge it to `main` firstly and then port to `v10.[version]` release branch.

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).

### CODE REVIEW

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt [hunk 1]
```diff
@@ -177,14 +177,18 @@ class CameraAnimationsPluginImpl : CameraAnimationsPlugin {
...
   @VisibleForTesting(otherwise = PRIVATE)
   internal fun performMapJump(cameraOptions: CameraOptions) {
-    if (lastCameraOptions == cameraOptions) {
+    if (lastCameraOptions == cameraOptions || cameraOptions.isEmpty()) {
       return
     }
     // move native map to new position
     try {
       mapCameraManagerDelegate.setCamera(cameraOptions)
+      logE("KIRYLDD", "setCamera: $cameraOptions")
```
Comment (alexander-kulikovskii) [inline plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:191]: rm? or set tag name

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt [hunk 2]
```diff
@@ -131,12 +139,53 @@ internal class ViewAnnotationManagerImpl(
...
-    drawAnnotationViews(positions)
+    // update that flag here if callback was triggered, it will be reset by renderer when swapping buffers
+    renderThread.hasViewAnnotations = true
+    // it's fine to update translation for views on non-main thread we simply update fields and let
+    // Android render node pick them up using Android Render thread. We don't even bother if view
+    // was not actually added on top of MapView.
+    positions.forEach { descriptor ->
+      annotationMap[descriptor.identifier]?.let { annotation ->
+        annotation.view.apply {
+          logE("KIRYLDD", "Translation upd post, mode = ${mode.name}")
```
Comment (alexander-kulikovskii) [inline sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:161]: tag name

File: sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationManager.kt [hunk 3]
```diff
@@ -144,4 +144,19 @@ interface ViewAnnotationManager {
...
+
+  /**
+   * Set view annotation update mode defined in [ViewAnnotationUpdateMode].
+   */
+  fun setViewAnnotationUpdateMode(mode: ViewAnnotationUpdateMode)
+
+  /**
+   * Static methods and variables.
+   */
+  companion object {
```
Comment (alexander-kulikovskii) [inline sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationManager.kt:156]: `internal` ?

File: sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt [hunk 4]
```diff
@@ -253,15 +263,53 @@ internal class MapboxRenderThread : Choreographer.FrameCallback {
...
+      renderTimeNsCopy = renderTimeNsCopy,
+      expectedEndRenderTimeNs = expectedEndRenderTimeNs
+    )
+  }
+
+  private fun swap(
+    nextFrameTimeNanos: Long,
+    currentFrameTimeNanos: Long,
+    renderTimeNsCopy: Long,
+    expectedEndRenderTimeNs: Long
```
Comment (alexander-kulikovskii) [inline sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt:302]:
```suggestion
    expectedEndRenderTimeNanos: Long
```

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt (hunk 1 above)
Comment (kiryldz) [reply plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:191]: leave for now, will remove later

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt (hunk 2 above)
Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt]: leave for now, will remove later

File: sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationManager.kt (hunk 3 above)
Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationManager.kt:156]: it's useful to expose our default value to the users so better leave public

File: sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt (hunk 4 above)
Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt:302]: done

File: sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationUpdateMode.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+   */
+  FIXED_DELAY,
+
+  /**
+   * View annotations are updated independent from MapView.
+   * View annotation position could be updated on the same frame as the map camera as well as on next one.
+   *
+   * Using this mode is not recommended as it will most likely bring in view annotation jittering.
+   */
+  MAP_INDEPENDENT,
```
Comment (kiryldz) [inline sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationUpdateMode.kt:24]: @mapbox/maps-android - to discuss if we need it at all. Looking at how it actually behaves I'm leaning to actually cut this enum out but would like to hear your opinion as well

Comment (pengdev) [reply sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationUpdateMode.kt:24]: Does this mode provide advantages compared with the other two modes? From the video this mode seems to be both laggy and jittering at the same time 🙈

Comment (pengdev) [reply sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationUpdateMode.kt:24]: I'm 👍 with removing if there's no specific advantages.

**Comment by yunikkk** [comment]
Would love to see the visual comparison of the same animations executed programmatically before/after to fully see the difference. Frame stats would be also very helpful

File: sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationUpdateMode.kt (hunk 5 above)
Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationUpdateMode.kt]: Did remove it indeed, I can't see any future for this mode with current core implementation. Thanks for your opinion @pengdev!

File: sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt [hunk 6]
```diff
@@ -253,15 +263,53 @@ internal class MapboxRenderThread : Choreographer.FrameCallback {
...
       mapboxRenderer.render()
+      logE("KIRYLDD", "render end $frameTimeNanos")
     }
 
     // assuming render event queue holds user's runnables with OpenGL ES commands
     // it makes sense to execute them after drawing a map but before swapping buffers
     // **note** this queue also holds snapshot tasks
     drainQueue(renderEventQueue)
+    if (hasViewAnnotations) {
+      hasViewAnnotations = false
```
Comment (yunikkk) [inline sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt:276]: Feels weird that after swapping buffers flag becomes false, I'd rename to `needViewAnnotationsSync` or smth

File: sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt [hunk 7]
```diff
@@ -253,15 +263,53 @@ internal class MapboxRenderThread : Choreographer.FrameCallback {
...
+            nextFrameTimeNanos = it,
+            currentFrameTimeNanos = frameTimeNanos,
+            renderTimeNsCopy = renderTimeNsCopy,
+            expectedEndRenderTimeNanos = expectedEndRenderTimeNanos
+          )
+        }
+        return
+      }
+    }
+    // perform swap immediately if no view annotations are visible or mode is not MAP_SYNCHRONIZED
```
Comment (yunikkk) [inline sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt:289]: Imho another comment is needed for `Choreographer.getInstance().postFrameCallback`, really worth to explain the subtle differences here

Also, can be simplified e.g.
`if (hasViewAnnotations && viewAnnotationMode == ViewAnnotationUpdateMode.MAP_SYNCHRONIZED) {} else {}`

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt [hunk 8]
```diff
@@ -177,14 +177,18 @@ class CameraAnimationsPluginImpl : CameraAnimationsPlugin {
     handler.removeCallbacks(commitChangesRunnable)
   }
 
+  private fun CameraOptions.isEmpty() =
+    center == null && bearing == null && pitch == null && padding == null && zoom == null && anchor == null
```
Comment (alexander-kulikovskii) [inline plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:181]: how not to forget update this method? maybe move to `CameraOptions.kt`?

Comment (kiryldz) [reply plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:181]: yep

File: sdk/src/test/java/com/mapbox/maps/ViewAnnotationManagerTest.kt [hunk 9]
```diff
@@ -305,6 +311,28 @@ class ViewAnnotationManagerTest {
...
+
+  @Test
+  fun setViewAnnotationUpdateMode() {
+    viewAnnotationManager.setViewAnnotationUpdateMode(ViewAnnotationUpdateMode.MAP_FIXED_DELAY)
+    verify(exactly = 1) { renderer.viewAnnotationMode = ViewAnnotationUpdateMode.MAP_FIXED_DELAY }
+  }
+
+  @Test
+  fun getViewAnnotationUpdateMode() {
+    viewAnnotationManager.setViewAnnotationUpdateMode(ViewAnnotationUpdateMode.MAP_FIXED_DELAY)
```
Comment (alexander-kulikovskii) [inline sdk/src/test/java/com/mapbox/maps/ViewAnnotationManagerTest.kt:331]: not fully understand why did you call `setViewAnnotationUpdateMode` while checking getter

File: sdk/src/test/java/com/mapbox/maps/renderer/MapboxRenderThreadTest.kt [hunk 10]
```diff
@@ -239,6 +241,57 @@ class MapboxRenderThreadTest {
     }
   }
 
+  @Test
+  fun needViewAnnotationSyncMapSynchronized() {
+    initRenderThread()
+    provideValidSurface()
+    // static method so does not matter which thread is called from -
+    // we need this interval for our render thread
+    ShadowChoreographer.setPostFrameCallbackDelay(16)
```
Comment (alexander-kulikovskii) [inline sdk/src/test/java/com/mapbox/maps/renderer/MapboxRenderThreadTest.kt:250]: is it possible that value for `setPostFrameCallbackDelay` and `idleHandler` can be different?
if yes, should you check it?

Comment (kiryldz) [reply sdk/src/test/java/com/mapbox/maps/renderer/MapboxRenderThreadTest.kt]: no, moved to variable for better readability

File: sdk/src/test/java/com/mapbox/maps/ViewAnnotationManagerTest.kt (hunk 9 above)
Comment (kiryldz) [reply sdk/src/test/java/com/mapbox/maps/ViewAnnotationManagerTest.kt]: thanks, revisited

**Review by alexander-kulikovskii — APPROVED** [review]

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt [hunk 11]
```diff
@@ -179,7 +180,9 @@ class CameraAnimationsPluginImpl : CameraAnimationsPlugin {
 
   @VisibleForTesting(otherwise = PRIVATE)
   internal fun performMapJump(cameraOptions: CameraOptions) {
-    if (lastCameraOptions == cameraOptions) {
+    // no need to call native setCamera if camera did not change or
+    // empty camera object has arrived (could be the case for gestures)
+    if (lastCameraOptions == cameraOptions || cameraOptions.isEmpty()) {
```
Comment (yunikkk) [inline plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:185]: is it somehow related to view annotations?

File: sdk/src/androidTest/java/com/mapbox/maps/ViewAnnotationTest.kt [hunk 12]
```diff
@@ -317,8 +320,6 @@ class ViewAnnotationTest(
         assertEquals(mapboxMap.pixelForCoordinate(CAMERA_CENTER).y, secondView.translationY.toDouble(), ADMISSIBLE_ERROR_PX)
         assertArrayEquals(
           arrayOf(
-            Pair(firstView, true),
```
Comment (yunikkk) [inline sdk/src/androidTest/java/com/mapbox/maps/ViewAnnotationTest.kt:320]: Why does the actualVisibilityUpdateList changes here?

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationBasicAddActivity.kt [hunk 13]
```diff
@@ -5,38 +5,36 @@ import android.os.Bundle
...
 
   override fun onCreate(savedInstanceState: Bundle?) {
     super.onCreate(savedInstanceState)
     val binding = ActivityViewAnnotationShowcaseBinding.inflate(layoutInflater)
     setContentView(binding.root)
 
-    viewAnnotationManager = binding.mapView.viewAnnotationManager
-
-    mapboxMap = binding.mapView.getMapboxMap().apply {
+    mapboxMap = mapView.getMapboxMap().apply {
```
[the remark this answers is not in the document]
Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/ViewAnnotationBasicAddActivity.kt:31]: yes, in xml

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt [hunk 14]
```diff
@@ -131,8 +136,41 @@ internal class ViewAnnotationManagerImpl(
...
+    renderThread.viewAnnotationMode = mode
+  }
+
+  @AnyThread
+  override fun getViewAnnotationUpdateMode(): ViewAnnotationUpdateMode {
+    return renderThread.viewAnnotationMode
+  }
+
+  /**
+   * First call will be always from Mapbox render thread that will be ALWAYS followed up
```
Comment (yunikkk) [inline sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:150]: I'd rephrased & simplified a bit, to highlight that
- we will have two calls with the same values from render thread, then from the main thread
- it's needed since scheduling to main thread from c++ is faster

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt [hunk 15]
```diff
@@ -131,8 +136,41 @@ internal class ViewAnnotationManagerImpl(
...
+   * another call from Android main thread.
+   */
+  @AnyThread
   override fun onViewAnnotationPositionsUpdate(positions: MutableList<ViewAnnotationPositionDescriptor>) {
-    drawAnnotationViews(positions)
+    // When called from render thread it means we're in the end of [MapInterface.render] call if positions did change.
```
Comment (yunikkk) [inline sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:155]: Here I'd just left indication which if branch happens first, and which goes second

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt (hunk 11 above)
Comment (kiryldz) [reply plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:185]: no, will remove and create follow-up ticket

File: sdk/src/androidTest/java/com/mapbox/maps/ViewAnnotationTest.kt (hunk 12 above)
Comment (kiryldz) [reply sdk/src/androidTest/java/com/mapbox/maps/ViewAnnotationTest.kt:320]: with previous a bit chaotic algo with events coming from different observers we actually did add the view and immediately remove it actually resulting the listeners being hit and make view annotation show up - although I guess end user did not see that as view annotation was added for a moment behind the map (z-index).

now due to more systemic  approach we simply add one view annotation (second one as it makes the first one disappear).

File: sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt [hunk 16]
```diff
@@ -261,8 +272,43 @@ internal class MapboxRenderThread : Choreographer.FrameCallback {
...
+      expectedVsyncWakeTimeNs = expectedEndRenderTimeNs - ONE_MILLISECOND_NS
+    }
+    fpsChangedListener?.let {
+      val fps = 1E9 / (actualEndRenderTimeNanos - timeElapsed)
+      if (timeElapsed != 0L) {
+        it.onFpsChanged(fps)
+      }
+      timeElapsed = actualEndRenderTimeNanos
+    }
+    if (needViewAnnotationSync && viewAnnotationMode == ViewAnnotationUpdateMode.MAP_SYNCHRONIZED) {
```
Comment (yunikkk) [inline sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt:289]: wdyt about introducing boolean property that checks both of them, and naming it like `swapBuffersOnNextVsync` or similar

Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt:289]: see no sense as it's checked in a single place and is more easy at least for me to understand then one more extra bool flag

File: sdk/src/test/java/com/mapbox/maps/renderer/MapboxRenderThreadTest.kt [hunk 17]
```diff
@@ -239,6 +241,59 @@ class MapboxRenderThreadTest {
...
+    // we need this interval for our render thread
+    ShadowChoreographer.setPostFrameCallbackDelay(choreographerCallbackDelayMs.toInt())
+    pauseHandler()
+    mapboxRenderThread.viewAnnotationMode = ViewAnnotationUpdateMode.MAP_SYNCHRONIZED
+    mapboxRenderThread.needViewAnnotationSync = true
+    mapboxRenderThread.queueRenderEvent(MapboxRenderer.repaintRenderEvent)
+    idleHandler(choreographerCallbackDelayMs)
+    // buffers are already swapped once when surface was created
+    // they should not be swapped this frame N, they need to be swapped next frame N+1
+    verify(exactly = 1) {
```
Comment (yunikkk) [inline sdk/src/test/java/com/mapbox/maps/renderer/MapboxRenderThreadTest.kt:259]:
```suggestion
    verifyOnce {
```

**Review by yunikkk — APPROVED** [review]

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/BaseViewAnnotationActivity.kt [hunk 18]
```diff
@@ -0,0 +1,32 @@
+package com.mapbox.maps.testapp.examples.markersandcallouts
+
+import android.view.Menu
+import android.view.MenuItem
+import androidx.appcompat.app.AppCompatActivity
+import com.mapbox.maps.testapp.R
+import com.mapbox.maps.viewannotation.ViewAnnotationManager
+import com.mapbox.maps.viewannotation.ViewAnnotationUpdateMode
+
+abstract class BaseViewAnnotationActivity : AppCompatActivity() {
```
Comment (pengdev) [inline app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/BaseViewAnnotationActivity.kt:10]: I see the point of introducing this base activity, since we have 3 view annotation related activities, but as examples, maybe each activity should be more self-contained, and it's easier to publish these examples to the `www` later as well.

File: sdk-base/api/metalava.txt [hunk 19]
```diff
@@ -59,6 +59,7 @@ package com.mapbox.maps.dsl {
   public final class CameraOptionsKtxKt {
     method public static inline com.mapbox.maps.CameraOptions cameraOptions(kotlin.jvm.functions.Function1<? super com.mapbox.maps.CameraOptions.Builder,kotlin.Unit> block);
     method public static inline com.mapbox.maps.CameraOptions cameraOptions(com.mapbox.maps.CameraState cameraState, kotlin.jvm.functions.Function1<? super com.mapbox.maps.CameraOptions.Builder,kotlin.Unit> block);
+    method public static boolean isEmpty(com.mapbox.maps.CameraOptions);
```
Comment (pengdev) [inline sdk-base/api/metalava.txt:62]: is this supposed to be a public API? or should it be an internal API only?

File: sdk-base/src/main/java/com/mapbox/maps/dsl/CameraOptionsKtx.kt [hunk 20]
```diff
@@ -24,4 +24,10 @@ inline fun cameraOptions(
...
   .apply(block)
-  .build()
\ No newline at end of file
+  .build()
+
+/**
+ * Check if camera options has all the empty properties.
+ */
+fun CameraOptions.isEmpty() =
+  center == null && bearing == null && pitch == null && padding == null && zoom == null && anchor == null
```
Comment (pengdev) [inline sdk-base/src/main/java/com/mapbox/maps/dsl/CameraOptionsKtx.kt:33]: does this need to be a public API?

File: sdk-base/api/metalava.txt (hunk 19 above)
Comment (kiryldz) [reply sdk-base/api/metalava.txt:62]: will be removed per discussion with @yunikkk - I'll create follow-up ticket later

File: sdk-base/src/main/java/com/mapbox/maps/dsl/CameraOptionsKtx.kt (hunk 20 above)
Comment (kiryldz) [reply sdk-base/src/main/java/com/mapbox/maps/dsl/CameraOptionsKtx.kt:33]: will be removed per discussion with @yunikkk - I'll create follow-up ticket later

File: app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/BaseViewAnnotationActivity.kt (hunk 18 above)
Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/markersandcallouts/BaseViewAnnotationActivity.kt:10]: well...
guess you're correct, will revert then

File: sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationManager.kt [hunk 21]
```diff
@@ -144,4 +144,24 @@ interface ViewAnnotationManager {
...
+  fun getViewAnnotationUpdateMode(): ViewAnnotationUpdateMode
+
+  /**
+   * Static methods and variables.
+   */
+  companion object {
+    /**
+     * Default view annotation update mode.
+     */
+    val DEFAULT_UPDATE_MODE = ViewAnnotationUpdateMode.MAP_SYNCHRONIZED
```
Comment (pengdev) [inline sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationManager.kt:165]: nit: annotate with JvmField.

File: sdk/api/metalava.txt [hunk 22]
```diff
@@ -582,15 +582,28 @@ package com.mapbox.maps.viewannotation {
...
+    method public com.mapbox.maps.viewannotation.ViewAnnotationUpdateMode getViewAnnotationUpdateMode();
     method public void removeAllViewAnnotations();
     method public void removeOnViewAnnotationUpdatedListener(com.mapbox.maps.viewannotation.OnViewAnnotationUpdatedListener listener);
     method public boolean removeViewAnnotation(android.view.View view);
+    method public void setViewAnnotationUpdateMode(com.mapbox.maps.viewannotation.ViewAnnotationUpdateMode mode);
     method public boolean updateViewAnnotation(android.view.View view, com.mapbox.maps.ViewAnnotationOptions options);
+    field public static final com.mapbox.maps.viewannotation.ViewAnnotationManager.Companion Companion;
+  }
+
+  public static final class ViewAnnotationManager.Companion {
```
Comment (pengdev) [inline sdk/api/metalava.txt:594]: we can get rid of companion by adding `JvmField` annotation to the `DEFAULT_UPDATE_MODE`

File: sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationUpdateMode.kt [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+  MAP_FIXED_DELAY,
+
+  /**
+   * View annotations are updated on the same frame as the MapView camera.
+   *
+   * Using this mode results in updating both map camera and view annotation position with 1-frame delay.
+   * Please note that using this mode may slightly decrease FPS when view annotations are visible on the map.
+   */
+  MAP_SYNCHRONIZED,
+}
```
Comment (pengdev) [inline sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationUpdateMode.kt:25]: nit: should we make it more extendable and use the pattern we talked about in `ViewportStatusChangeReason`?

File: sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt [hunk 24]
```diff
@@ -131,8 +136,41 @@ internal class ViewAnnotationManagerImpl(
...
+    // It's crucial to notify render thread in this callback as depending on mode we're using we need
+    // either swap buffers the same or the next frame.
+    if (Looper.myLooper() == Looper.getMainLooper()) {
+      // create copy to avoid concurrent modification exception
+      val immutablePositionListCopy = LinkedList(updatedPositionsList)
```
[the remark this answers is not in the document]
Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt:160]: yep, indeed

File: sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationUpdateMode.kt (hunk 23 above)
Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationUpdateMode.kt:25]: nice idea, will do!

**Review by pengdev — APPROVED** [review]
LGTM, please also update the PR description

**Review by yunikkk — APPROVED** [review]
